src/theory/datatypes/theory_datatypes_type_rules.h

```diff
@@ -427,7 +427,115 @@ class DtSyguEvalTypeRule
     }
     return TypeNode::fromType(dt.getSygusType());
   }
-}; /* class DtSygusBoundTypeRule */
+}; /* class DtSyguEvalTypeRule */
+
+class MatchTypeRule
+{
+ public:
+  inline static TypeNode computeType(NodeManager* nodeManager,
+                                     TNode n,
+                                     bool check)
```

Comment:
We can avoid a few things, I've added a guard to block some of the work.  We still need to get the type of each child since the return type is the least common type of the cases.